rasa/engine/validation.py

```diff
@@ -0,0 +1,324 @@
+import inspect
+import logging
+import typing
+from typing import Optional, Callable, Text, Tuple, Dict, Type, Any, Set
+
+import dataclasses
+import typing_utils
+
+from rasa.engine.exceptions import GraphSchemaValidationException
+from rasa.engine.graph import (
+    GraphSchema,
+    GraphComponent,
+    SchemaNode,
+    Fingerprintable,
+    ExecutionContext,
+)
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.nlu import components
+
+
+@dataclasses.dataclass
+class ParameterInfo:
+    """Stores metadata about a function parameter."""
+
+    type_annotation: Any
+    is_kwargs: bool
+    has_default: bool
+
+
+_keywords_expected_types: Dict[Text, Any] = {
+    "resource": Resource,
+    "execution_context": ExecutionContext,
+    "model_storage": ModelStorage,
+    "config": Dict[Text, Any],
+}
+
+
+def validate(
+    schema: GraphSchema, language: Optional[Text], is_train_graph: bool
+) -> None:
+    """Validates a graph schema.
+
+    This checks that the graph structure is correct (e.g. all nodes pass the correct
+    things into each other) and validates the single graph components.
+
+    Args:
+        schema: The schema which needs validating.
+        language: Used to validate if all components support the language the assistant
+            is used in. If the language is `None`, all components are assumed to be
+            compatible.
+        is_train_graph: Whether the graph is used for training.
+
+    Raises:
+        GraphSchemaValidationException: If the validation failed.
+    """
+    for node_name, node in schema.nodes.items():
+        _validate_interface_usage(node_name, node)
+        _validate_supported_languages(language, node, node_name)
+        _validate_required_packages(node, node_name)
+
+        run_fn_params, run_fn_return_type = _get_parameter_information(
+            node_name, node.uses, node.fn
+        )
+        _validate_run_fn(
+            node_name, node, run_fn_params, run_fn_return_type, is_train_graph
+        )
+
+        create_fn_params, _ = _get_parameter_information(
+            node_name, node.uses, node.constructor_name
+        )
+        _validate_constructor(node_name, node, create_fn_params)
+
+        _validate_needs(
+            node_name, node, schema, create_fn_params, run_fn_params,
+        )
+
+
+def _validate_interface_usage(node_name: Text, node: SchemaNode) -> None:
+    if not issubclass(node.uses, GraphComponent):
```

Comment:
Unfortunately doesn't look at types or we wouldn't have to check the keyword parameters later...

Reply:
Well, as the `run fn` is somewhat arbitrary as well as the `loads` signature the protocol check would never really help us 😬